Harden AssemblyDependencyResolver assemblyPaths

LakshanF opened the pull request:
Pick the first simple assembly for multiples

AssemblyDependencyResolver is made resilience to the case when 
hostpolicy.dll returns multiple assembly paths for the same assembly 
by picking the first one.

Fix #37162

LakshanF committed:
Harden AssemblyDependencyResolver assemblyPaths  AssemblyDependencyResolver is made resilience to the case where  hostpolicy.dll returns the same assembly paths by ignoring multiple  entries. This is done by using OrdinalIgnoreCase comparison on Windows and Ordinal comparison elsewhere  Fix #37162

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyDependencyResolver.cs:
```diff
@@ -95,7 +105,16 @@ public AssemblyDependencyResolver(string componentAssemblyPath)
             {
-                _assemblyPaths.Add(Path.GetFileNameWithoutExtension(assemblyPath), assemblyPath);
+                // Check to see if we have the same managed assembly simple name in the dictionary.
+                // We compare the paths reported and if they're the same, simply ignore the duplicate entries. And only fail if the paths are different
```
Are there user errors when we expect this to fail? Would it make sense to simplify it to just `_assemblyPaths.TryAdd(Path.GetFileNameWithoutExtension(assemblyPath), assemblyPath);` ?

LakshanF replied:
Yes, my understanding is that we want to continue to throw when the same simple name is used but not for the same assembly with the same full path. We decided to fix this in ADR rather than in the native code given the challenges in getting the casing right

vitek-karas replied:
Yes - it's possible to run into weird cases where this should fail. One such example:
* On Linux I have a Plugin.dll and Plugin.deps.json
* I create a plugin.dll in the same directory and rename Plugin.deps.json to plugin.deps.json
* I use ADR on plugin.dll

This will result in hostpolicy potentially reporting both Plugin.dll and plugin.dll paths - and we don't know which one to pick for the dictionary - so it should fail.

jkotas replied:
Can the same situation and failure mode happen for regular executables?

vitek-karas replied:
Maybe - we should try that

LakshanF replied:
Decided to stay away from case comparison and take your original suggestion for the fix

jkotas commented:
>  challenges in getting the casing right

Note that Windows supports case sensitive file systems (and vice versa on Linux). Comparing paths like what is done here is close to impossible to get 100% right.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyDependencyResolver.cs:
```diff
@@ -95,7 +105,16 @@ public AssemblyDependencyResolver(string componentAssemblyPath)
+                // On Windows - OrdinalIgnoreCase comparison
+                // Elsewhere - Ordinal comparison
+                string assemblySimpleName = Path.GetFileNameWithoutExtension(assemblyPath);
+                if (!_assemblyPaths.ContainsKey(assemblySimpleName) || !_assemblyPaths[assemblySimpleName].Equals(assemblyPath, StringComparisonMode))
```
`ContainsKey` here is redundant lookup for the common case. If we decide that we want to have some error handling for collisions, it would be better to call `TryAdd` and only do the additional error handling work if there is a collision:

```
if (!_assemblyPaths.TryAdd(Path.GetFileNameWithoutExtension(assemblyPath), assemblyPath))
{
    decide whether to throw exception or not
}
```

LakshanF replied:
I'll chat with @vitek-karas more to understand the existing tests and to see if we can add a case to cover this scenario

LakshanF replied:
Added tests and also opened issue #42334 to track some behavioral differences between mac-Linux host policy

LakshanF committed:
Pick the first simple assembly for multiples

AssemblyDependencyResolver is made resilience to the case when
hostpolicy.dll returns multiple assembly paths for the same assembly
by picking the first one.

Fix #37162

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyDependencyResolver.cs:
```diff
@@ -27,6 +27,7 @@ public sealed class AssemblyDependencyResolver
         private readonly string[] _resourceSearchPaths;
         private readonly string[] _assemblyDirectorySearchPaths;
 
+
```
Nit: extra new line

LakshanF committed:
missed an extra line

jkotas commented:
How hard is it to add a test for the problem reported in the issue?

vitek-karas reviewed (approved):
We should look into testing, but the product change looks good.

LakshanF committed:
Test case for casing change with the same assembly

LakshanF committed:
typo

LakshanF committed:
Disabled the tests for non-windows

Filed #42334 to track the linux-mac differences

LakshanF commented on src/installer/tests/HostActivation.Tests/DependencyResolution/ResolveComponentDependencies.cs:
```diff
@@ -49,6 +49,41 @@ public void ComponentWithNoDependenciesAndNoDeps()
+
+            // Rename
+            File.Move(fileName, changeFile);
+            File.Delete(component.DepsJson);
```
Added this scenario and also opened issue #42334 to track some behavioral differences between mac-Linux host policy

agocke commented:
/backport release/5.0

agocke commented:
/backport to release/5.0

agocke commented:
@akoeplinger Any chance you know why the backport command isn't working?

agocke commented:
/backport to release/5.0

agocke commented:
nvm, it was because it was convo locked.
